content/languages/cpp/authoring/index.md

```diff
@@ -0,0 +1,388 @@
+
+```cpp
+template<typename ActualType, typename ConstraintListType, typename MessageSupplierType>
+static void That(const ActualType& actual, ExpressionBuilder<ConstraintListType> expression, const MessageSupplierType& message_supplier, const char* file_name = "", int line_number = 0);
+
+template<typename ConstraintListType, typename MessageSupplierType>
+static void That(const char* actual, ExpressionBuilder<ConstraintListType> expression, const MessageSupplierType& message_supplier, const char* file_name = "", int line_number = 0)
+    
+template<typename ActualType, typename ExpressionType, typename MessageSupplierType>
+static void That(const ActualType& actual, const ExpressionType& expression, const MessageSupplierType& message_supplier, const char* file_name = "", int line_number = 0);
+
+template<typename ExpressionType, typename MessageSupplierType>
+static void That(const char* actual, const ExpressionType& expression, const MessageSupplierType& message_supplier, const char* file_name = "", int line_number = 0);
+```
+
+`message_supplier` is a callable compatible with a function accepting no arguments and returning `std::string`. It can be a function, a functor, a lambda expression, or an instance of any type supporting above requirements.
+
+For more details and examples of custom assertion messages, see `[Example test suite](#example-test-suite)` below, or visit the `[Snowhouse reference](/languages/cpp/snowhouse/)` page.
+
+#### Stringizers
+
+In its default configuration, Snowhouse can produce confusing assertion messages when `expected` and `actual` values are of a type it cannot stringify:
+
+```text
+  does_not_pretty_print_type_without_stringizer
+    Expected: equal to [ [unsupported type], [unsupported type] ]
+    Actual: [ [unsupported type], [unsupported type] ]
+```
+
+To rectify this issue in your tests, you can make such types suitable for stringification in a way described in the dedicated documentation article on [custom Snowhouse stringizers](/languages/cpp/igloo/stringizers/).
+
+
+### Random utilities
+
+Until C++11, the most common way of producing random values was the `rand` function. However, it has a set of problems: it needs to be properly seeded, and it is difficult to produce values outside of `0...RAND_MAX` range. Since C++11, the standard library offers a set of functionalities which are designed to produce random values of different types, from various ranges, and with better distribution when compared to `rand`. Unfortunately, the API presented in the `random` header seems to be confusing and difficult to use and accompanied by a few misconceptions, and authors either use it incorrectly or resort to (not) good, old `rand`. However, working with `random` turns out to be not that difficult:
+
+```cpp
+//use random_device only as a seeder
```

Comment:
Can we add a space after `//`? It's difficult to read.

I also think it's more common to have a space in any language.

```javascript
// comment
//comment
``` 
```python
# comment Python mentions this in PEP8 too.
#comment
```
```haskell
-- comment
--comment
```
```ocaml
(* comment *)
(*comment*)
```

Reply:
I think it's difficult for me to read because my brain tries to ignore it. I've only seen no space when commenting out the line.

For the following code:
```javascript
//foo
bar();
```
I read:
```javascript
bar();
```
until I try not to ignore.

That doesn't happen when it has a space:
```javascript
// foo
bar();
```

Reply:
Found https://stackoverflow.com/questions/1467058/space-between-line-comment-characters-and-start-of-actual-comment and the answers matches my experience.

I'll make it a rule. In our docs, space is required.

Reply:
Done.